clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -369,13 +459,18 @@ private void handleSaslToken(byte[] clientToken) throws IOException {
                 // This should not normally occur since clients typically check supported versions using ApiVersionsRequest
                 throw new UnsupportedVersionException("Version " + version + " is not supported for apiKey " + apiKey);
             }
+            if (!connectedClientSupportsReauthentication)
```

Comment:
The client sends multiple `SASL_AUTHENTICATE` requests, and I was thinking the client is known to support the latest version if it sends at least one of them with the required version.  I realize it is unlikely to send different versions each time, but technically if it sends the required version just once then we know it supports that version.  So I was trying to take that (admittedly rare) possibility into account.  Another way to do it would be:
```
reauthInfo.connectedClientSupportsReauthentication = reauthInfo.connectedClientSupportsReauthentication
        || version > 0;
```
I figured the way I did it was better than that since it avoids the write when the value is already true.

I can always set it if you feel that is more appropriate -- just let me know, otherwise I'll leave it as-is.

Reply:
In the meantime, I added a comment describing the reasoning.